text/3437-implementable-trait-alias.md

```diff
@@ -0,0 +1,319 @@
+
+An impl block for a trait alias looks just like an impl block for the underlying trait. The alias's where clauses
+are treated as obligations that the the `impl`ing type must meet.
+
+```rust
+pub trait CopyIterator = Iterator<Item: Copy> where Self: Send;
+
+struct Foo;
+
+impl CopyIterator for Foo {
+    type Item = i32; // Would be an error if this was `String`
+
+    fn next(&mut self) -> Self::Item {
+        42
+    }
+}
+
+struct Bar;
+impl !Send for Bar;
+
+// ERROR: `Bar` is not `Send`
+// impl IntIterator for Bar { /* ... */ }
+```
+
+If the trait alias uniquely constrains a portion of the `impl` block, that part can be omitted.
+
+```rust
+pub trait IntIterator = Iterator<Item = i32> where Self: Send;
+
+struct Baz;
+
+impl IntIterator for Baz {
+    // The alias constrains `Self::Item` to `i32`, so we don't need to specify it
+    // (though we are allowed to do so if desired).
+    // type Item = i32;
+
+    fn next(&mut self) -> i32 {
+        -27
+    }
+}
+```
+
+Alias `impl`s also allow omitting implied `#[refine]`s:
+
+```rust
+//! crate frob-lib
+#![feature(trait_alias)]
+
+pub trait LocalFrobber {
+    async fn frob(&self);
+}
+
+// not `+ Send`!
+pub trait Frobber = LocalFrobber<frob(..): Send> where Self: Send;
+```
+
+```rust
+//! crate joes-crate
+#![feature(trait_alias_impl)]
+
+use frob_lib::Frobber;
+
+struct MyType;
+
+impl Frobber for MyType {
+    // The return future of this method is implicitly `Send`, as implied by the alias.
+    // No `#[refine]` is necessary.
+    async fn frob(&self) {
+        println!("Sloo is 120% klutzed. Initiating brop sequence...")
+    }
+}
+```
+
+Trait aliases are `unsafe` to implement iff the underlying trait is marked `unsafe`.
+
+# Drawbacks
+
+- The sytactic distance between implementable and non-implementable aliases is short, which might confuse users.
+In particular, the fact that `trait Foo = Bar + Send;` means something different than `trait Foo = Bar where Self: Send;` will likely be surprising to many.
+- Adds complexity to the language.
+- Many of the motivating use-cases involve language features that are not yet stable, or even merely speculative.
+More experience with those features might unearth better alternatives.
+
+# Rationale and alternatives
+
+- Very lightweight, with no new syntax forms. Compare "trait transformers" proposals, for example—they are generally much heavier.
+- Better ergonomics compared to purely proc-macro based solutions.
+- One alternative is to allow marker traits or auto traits to appear in `+` bounds of implementable aliases.
+(For example, `trait Foo = Bar + Send;` could be made implementable). However, I suspect that the complexity would not be worthwhile.
+- Another possibility is to require an attribute on implmenentable aliase; e.g. `#[implementable] trait Foo = ...`. Again, I don't think that the complexity is warranted.
```

Comment:
one benefit of requiring `#[implementable]` is that it mitigates the confusing aspect of:
```rust
#[implementable] // error: must only have one trait on rhs of equal -- generated because of #[implementable]
pub trait Foo = Bar + Baz;
// vs.
#[implementable]
pub trait Foo = Bar
where
    Self: Baz;
```

Reply:
I've added some more discussion about this.